In [ ]:
# !pip install geopy

In [2]:
import cv2
import os 
import numpy as np
import pandas as pd
try:
    from PIL import Image
except ImportError:
    import Image
import pytesseract
import matplotlib.pyplot as plt
from skimage import util 
from skimage.color import rgb2gray

# Get geo

In [23]:
import geopy

locator = geopy.Nominatim(user_agent='myGeocoder')
location = locator.geocode('Champ de Mars, Paris, France')
location = locator.geocode('Альметьевск; Белоглазова 131')
print('Latitude = {}, Longitude = {}'.format(location.latitude, location.longitude))

Latitude = 54.90249145, Longitude = 52.2897052


In [109]:
# тут ручками начали форматировать и получили "Геолокация камер формат.xlsx"
# df = pd.read_excel('../Геолокация камер формат.xlsx',index_col='ID папки')
# print(df.Type.value_counts())
# df['lat']=np.nan
# df['lon']=np.nan
# for i,adress in enumerate(df['Адрес установки камеры'].values):
#     try: 
#         df['lat'][i] = locator.geocode(adress).latitude
#         df['lon'][i] = locator.geocode(adress).longitude
#     except:
#         pass
# df.to_csv('../geo_cam.csv',encoding='cp1251',sep=';')
# тут ручками начали забивать
# df = pd.read_csv('../geo_cam.csv',encoding='cp1251',sep=';',index_col='ID папки')
# for (idx, val) in df['lon'][df['lon'].isna()].iteritems():
#     lat,lon =  df.loc[idx,'lat'].split(', ')
#     df.loc[idx,'lon'] = lon
#     df.loc[idx,'lat'] = lat
# df.to_csv('geo_cam_done.csv')

Адрес установки камеры Type  \
ID папки                                                     
0                        Альметьевск, Белоглазова 131  ТКО   
1                        Альметьевск, Белоглазова 151  ТКО   
2                             Альметьевск, Джалиля 15  ТКО   
3                       Альметьевск, Гафиатуллина 29Б  ТКО   
4                        Альметьевск, Гафиатуллина 39  ТКО   
5                        Альметьевск, Гафиатуллина 45  ТКО   
6          Альметьевск, Гафиатуллина 47 , подъезд  1   ТКО   
7            Альметьевск, Гафиатуллина 47, подъезд 2   ТКО   
8                       Альметьевск, Гафиатуллина 51Б  ТКО   
9                             Альметьевск, Ленина 109  ТКО   
10                             Альметьевск, Ленина 66  ТКО   
11                             Альметьевск, Ленина 90  ТКО   
12                           Альметьевск, Шевченко 80  ТКО   
13                        Альметьевск, Строителей 20А  ТКО   
14                        Альметьевск, Строителей 20Б  ТКО   
15                         Альметьевск, Строителей 20  ТКО   
16                                Казань, Блюхера 79   ТКО   
17                          Казань, Чистопольская 61А   БД   
18                                Казань, Химиков 19    БД   
19                 Казань, Кариева 4а к2, подъезд  1    БД   
20                 Казань, Кариева 4а к2, подъезд  2    БД   
21                    Казань, Родины 26Е, подъезд  1    БД   
22                     Казань, Родины 26Е, подъезд  2   БД   
23                   Казань, Симонова 15, подъезд  1    БД   
24                            Казань, ул. Чапаева 11    БД   
25                             Казань, ул. Чапаева 16   БД   
26                             Казань, ул. Чапаева 16   БД   
27             Казань, ул. Побежимова 36, подъезд  1    БД   
28              Казань, ул. Побежимова 36, подъезд  3   БД   
29              Казань, ул. Побежимова 36, подъезд  4   БД   
30              Казань, ул. Побежимова 36, подъезд  5   БД   
31              Казань, ул. Побежимова 36, подъезд  6   БД   
32              Казань, ул. Побежимова 36, подъезд  7   БД   
33             Казань, ул. Побежимова 36, подъезд  2    БД   
34                          Казань, ул. Побежимова 36   БД   
35                          Казань, ул. Побежимова 36   БД   
36                          Казань, ул. Побежимова 36   БД   
37                          Казань, ул. Побежимова 36   БД   
38                          Казань, ул. Побежимова 36   БД   
39                          Казань, ул. Побежимова 36   БД   
40        г. Казань, Проспект Победы д. 139, корпус 3   БД   
41                  г. Казань, ул. Вишневского, д. 49   БД   
42                      г. Казань, Даурская улица, 24   БГ   
43               г. Казань, улица Рабочей Молодёжи, 6   БГ   
44              г. Казань, улица Рабочей Молодёжи, 18   БГ   
45              г. Казань, улица Рабочей Молодёжи, 22   БГ   
46                       г. Казань, ул. Баумана, д.47   БГ   
47                     г. Казань, улица Богатырёва, 8   БГ   
48                  г. Казань, улица Рустема Яхина, 8   БГ   
49                       г. Казань ,улица Баумана, 36   БГ   
50                     г. Казань, улица Баумана, 9/11   БГ   
51                    г. Казань ул. Дементьева, д. 33   БГ   
52            г. Казань, территория Озеро Лебяжье, 1А   БГ   
53                   г. Казань ул. Кремлевская, д. 2    БГ   
54                      г. Казань, ул. Молодежная д.8   БГ   
55                г. Казань,улица Николая Ершова, 4/2   БГ   
56                        г. Казань, ул. Пушкина д.17   БГ   
57                   г. Казань, улица Гаяза Исхаки, 1   БГ   

                           lat        lon  
ID папки                                   
0                  54.90249145  52.289705  
1                  54.90021055  52.287217  
2                   54.8969668  52.296581  
3                   54.9055572  52.270927  
4                  54.90640845  52.266456  
5 

# Извлечение районов

In [124]:
.reverse("52.509669, 13.376294")

In [12]:
import geopy
locator = geopy.Nominatim(user_agent='myGeocoder')
location = locator.reverse("55.784364, 49.185636")

In [20]:
location.address

'4А к2, улица Габдуллы Кариева, ЖК «Казань ХХI век», Советский район, Казань, городской округ Казань, Татарстан, Приволжский федеральный округ, 420081, Россия'

Location(4А к2, улица Габдуллы Кариева, ЖК «Казань ХХI век», Советский район, Казань, городской округ Казань, Татарстан, Приволжский федеральный округ, 420081, Россия, (55.78433085, 49.185579679308375, 0.0))

Location(11, улица Чапаева, Кадышево, Авиастроительный район, Казань, городской округ Казань, Татарстан, Приволжский федеральный округ, 420096, Россия, (55.8970477, 49.129442657771975, 0.0))

In [22]:
df = pd.read_csv('geo_cam_done.csv',index_col='ID папки')
# for i in df.iterrows():
#     print(i)
df['Район'] = np.nan
df['Управл'] = np.nan
for i in df.index[16:]:
    location = locator.reverse(f'{df.lat.values[i]}, {df.lon.values[i]}').address.split(', ')
    for element in location:
        if 'райо' in element.lower():
            df.loc[i,'Район'] = element
df.to_csv('../geo_cam_rai.csv',encoding='cp1251',sep=';')
df
    
    

Адрес установки камеры Type        lat  \
ID папки                                                                
0                        Альметьевск, Белоглазова 131  ТКО  54.902491   
1                        Альметьевск, Белоглазова 151  ТКО  54.900211   
2                             Альметьевск, Джалиля 15  ТКО  54.896967   
3                       Альметьевск, Гафиатуллина 29Б  ТКО  54.905557   
4                        Альметьевск, Гафиатуллина 39  ТКО  54.906408   
5                        Альметьевск, Гафиатуллина 45  ТКО  54.905825   
6          Альметьевск, Гафиатуллина 47 , подъезд  1   ТКО  54.906021   
7            Альметьевск, Гафиатуллина 47, подъезд 2   ТКО  54.906161   
8                       Альметьевск, Гафиатуллина 51Б  ТКО  54.906745   
9                             Альметьевск, Ленина 109  ТКО  54.898048   
10                             Альметьевск, Ленина 66  ТКО  54.900030   
11                             Альметьевск, Ленина 90  ТКО  54.898638   
12                           Альметьевск, Шевченко 80  ТКО  54.896065   
13                        Альметьевск, Строителей 20А  ТКО  54.901198   
14                        Альметьевск, Строителей 20Б  ТКО  54.900797   
15                         Альметьевск, Строителей 20  ТКО  54.895480   
16                                Казань, Блюхера 79   ТКО  55.825624   
17                          Казань, Чистопольская 61А   БД  55.819753   
18                                Казань, Химиков 19    БД  55.867680   
19                 Казань, Кариева 4а к2, подъезд  1    БД  55.784263   
20                 Казань, Кариева 4а к2, подъезд  2    БД  55.784364   
21                    Казань, Родины 26Е, подъезд  1    БД  55.763124   
22                     Казань, Родины 26Е, подъезд  2   БД  55.763421   
23                   Казань, Симонова 15, подъезд  1    БД  55.865815   
24                            Казань, ул. Чапаева 11    БД  55.897048   
25                             Казань, ул. Чапаева 16   БД  55.863513   
26                             Казань, ул. Чапаева 16   БД  55.863294   
27             Казань, ул. Побежимова 36, подъезд  1    БД  55.860504   
28              Казань, ул. Побежимова 36, подъезд  3   БД  55.860288   
29              Казань, ул. Побежимова 36, подъезд  4   БД  55.860290   
30              Казань, ул. Побежимова 36, подъезд  5   БД  55.860342   
31              Казань, ул. Побежимова 36, подъезд  6   БД  55.860388   
32              Казань, ул. Побежимова 36, подъезд  7   БД  55.860480   
33             Казань, ул. Побежимова 36, подъезд  2    БД  55.860696   
34                          Казань, ул. Побежимова 36   БД  55.860353   
35                          Казань, ул. Побежимова 36   БД  55.860095   
36                          Казань, ул. Побежимова 36   БД  55.860160   
37                          Казань, ул. Побежимова 36   БД  55.859936   
38                          Казань, ул. Побежимова 36   БД  55.860629   
39                          Казань, ул. Побежимова 36   БД  55.859917   
40        г. Казань, Проспект Победы д. 139, корпус 3   БД  55.775768   
41                  г. Казань, ул. Вишневского, д. 49   БД  55.785174   
42                      г. Казань, Даурская улица, 24   БГ  55.763462   
43               г. Казань, улица Рабочей Молодёжи, 6   БГ  55.779575   
44              г. Казань, улица Рабочей Молодёжи, 18   БГ  55.780335   
45              г. Казань, улица Рабочей Молодёжи, 22   БГ  55.781023   
46                       г. Казань, ул. Баумана, д.47   БГ  55.788171   
47                     г. Казань, улица Богатырёва, 8   БГ  55.816106   
48                  г. Казань, улица Рустема Яхина, 8   БГ  55.789676   
49                       г. Казань ,улица Баумана, 36   БГ  55.793019   
50                     г. Казань, улица Баумана, 9/11   БГ  55.792594   
51                    г. Казань ул. Дементьева, д. 33   БГ  55.860105   
52            г. Казань, территория Озеро Лебяжье, 1А   БГ  55.837995   
53              

In [30]:
for i in pd.read_csv('../macro_micro_analytics/geo_cam_rai.csv',encoding='cp1251',sep=';',header=None)[6].unique():
    print(i)

УК «ЖКХ Московский район»
УК «УЮТНЫЙ ДОМ ГРУПП»
УК «Управляющая компания Райжилуправления Советского района»
УК «ЖиК»
УК «ЖКХ Авиастроительного района»
УК «СТАРЫЙ ГОРОД»
УК «Управляющая компания Вахитовского района»
УК «УЮТСЕРВИС»
УК «УПРАВЛЯЮЩАЯ КОМПАНИЯ РАН СЕРВИС»


# OCR

In [ ]:
pytesseract.pytesseract.tesseract_cmd = r'.\OCR\tesseract'

In [20]:
def preproc(video):
    """
    на вход название видео, в результате созадет 
    папку с этим названием и по кадрово ложит туда 
    600 на 600 картинки. 
    В результате возращаем массивы: широты, долготы, скорости, Таймстемпы пандоса, и собсвтенно эти картинки 600:600
    """
    name = video.split('/')[-1].replace('.MOV','')
    os.mkdir(name)
    vidcap = cv2.VideoCapture(video)
    success,image = vidcap.read()
    pytesseract.pytesseract.tesseract_cmd = r'.\OCR\tesseract'
    count = 0
    n=[]
    r=[]
    speed = []
    time = []
    test = []

    while success:
        cv2.imwrite(f"{name}/frame{count}.jpg" , cropCenter(image,(600,600)))     # save frame as JPEG file      

        #       print('Read a new frame: ', success)
        if success:
            try:
                x = pytesseract.image_to_string(util.invert(image[1000:,:1000])).replace('\n','').split() 
                n.append(float(x[0][2:]))
                r.append(float(x[1][2:]))
                speed.append(float(x[2]))
                time.append(pd.Timestamp(x[4].replace('/','.')+' '+x[5]))
                test.append(cropCenter(image,(600,600)))
                
            except:
                print(count,'error',x)
        
        count += 1
        success,image = vidcap.read()
    return np.array(n),np.array(r),np.array(speed),np.array(time),np.array(test)

In [10]:
# vidcap = cv2.VideoCapture('7.MOV')
# success,image = vidcap.read()

In [12]:
# plt.imshow(cropCenter(image,(600,600)))

In [21]:
# plt.figure(figsize=(15,4))
# plt.imshow(rgb2gray(util.invert(image[1000:,:1000])))
# pytesseract.image_to_string(util.invert(image[1000:,:1000]))

In [ ]:
preproc('7.MOV')

97 error ['N:59.29025', 'E£:32.20875', '30', 'KM/H', '24/09/2020', 'Tiegieet']
98 error ['N:59.29025', 'E£:32.20875', '30', 'KM/H', '24/09/2020', 'Tiegieet']
99 error ['N:59.29025', 'E£:32.20875', '30', 'KM/H', '24/09/2020', 'Tiegieet']
100 error ['N:59.29025', 'E£:32.20875', '30', 'KM/H', '24/09/2020', 'Tiegieet']
101 error ['N:59.29025', 'E£:32.20875', '30', 'KM/H', '24/09/2020', 'Tiegieet']
102 error ['N:59.29025', 'E£:32.20875', '30', 'KM/H', '24/09/2020"', 'hiegieet']
103 error ['N:59.29025', 'E£:32.20875', '30', 'KM/H', '24/09/2020)', 'TieZiee', 'tT']
104 error ['N:59.29025', 'E£:32.20875', '30', 'KM/H', '24/09/2020)', 'TieZiee', 'tT']
105 error ['N:59.29025', 'E£:32.20375', '30', 'KM/H', '24/09/2020"', 'Ti', 's2iew']
106 error ['N:59.29025', 'E£:32.20375', '30', 'KM/H', '24/09/2020"', 'Ti', '2ieew']
107 error ['N:59.29025', 'E£:32.20375', '30', 'KM/H', '24/09/2020"', 'Tiare', 'ew']
108 error ['N:59.29025', 'E£:32.20375', '30', 'KM/H', '24/09/2020"', 'Tiare', 'ew']
109 error ['N:

In [25]:

# Simple image to string


In [29]:
x= tt
x 

['N:59.28997', 'R:32.20857', '26', 'KM/H', '24/09/2020', '11:21:27', 'CARCAM']

In [33]:
np.datetime64('24.09.2020 11:21:27')

ValueError: Error parsing datetime string "24.09.2020 11:21:27" at position 2

In [36]:
np.datetime64(pd.Timestamp('24.09.2020 11:21:27'))

numpy.datetime64('2020-09-24T11:21:27.000000')

In [41]:
r

32.20857

In [42]:
n

59.28997